### PR TITLE
Syndicate painting teleporter which can lock on to any active syndicate teleporter beacons

### PR DESCRIPTION
**Pull request by Xander3359**
## About The Pull Request
1-
Adds a teleporter-disguised-as-a-painting which you can buy from an uplink.
This teleporter can lock on to any active syndicate teleporter, and allows 2-way teleportation between the painting and the teleporter.
2-
Does a few mapping changes to the syndicate lava base
- Adds a defib + mount to medical
- Slightly changes the dorms area to place a teleporter
- Adds more turrets to the lava base
- Adds more surrounding lava
## Why It's Good For The Game
See current design doc thread on the discord: https://discord.com/channels/326822144233439242/1233023738878885938
But TLDR this allows ghost role syndies to help agents in a more direct way rather than just mucking up the comms.
THIS IS NOT A FREE RESAPWN, agents cannot go through the teleporter.
## Changelog
:cl:
add: a painting teleporter that traitors can buy in their uplink
add: slight changes to the syndie lava base, including teleporter, defib, more lava and turrets

add: Added new mechanics or gameplay changes
add: Added more things
del: Removed old things
qol: made something easier to use
balance: rebalanced something
fix: fixed a few things
sound: added/modified/removed audio or sound effects
image: added/modified/removed some icons or images
spellcheck: fixed a few typos
code: changed some code
refactor: refactored some code
config: changed some config setting
admin: messed with admin stuff
server: something server ops should know
/:cl:

### CODE REVIEW

**Comment by sylvia-from-fulp-station** [comment]
this is very cool i want to use lavaland base as a secret hideout

File: code/game/machinery/teleporter.dm [hunk 1]
```diff
@@ -97,12 +97,67 @@
...
 
+/obj/machinery/teleport/hub/syndicate
+	///When a painting targets us as their signal, we save them as our return target
+	var/atom/return_target
+	///Internal timer to prevent audio spam.
+	var/next_beep = 0
+
+/obj/machinery/teleport/hub/syndicate/Destroy()
+	if(src in GLOB.active_syndicate_beacons)
+		GLOB.active_syndicate_beacons -= src
```
Comment (MrMelbert) [inline code/game/machinery/teleporter.dm:108]:
```suggestion
	GLOB.active_syndicate_beacons -= src
```
This works fine even if `src` is not in that list

**Comment by EricZilla** [comment]
1. Set up syndicate painting
2. Free respawns

File: code/game/machinery/teleporter.dm [hunk 2]
```diff
@@ -97,12 +97,67 @@
...
+		var/atom/old_return_target = return_target
+		return_target = null
+		SEND_SIGNAL(old_return_target, TELEPORTER_SET_TARGET, src)
+		update_appearance(UPDATE_ICON)
+
+/obj/machinery/teleport/hub/syndicate/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	if(!(src in GLOB.active_syndicate_beacons))
+		GLOB.active_syndicate_beacons += src
+		user.balloon_alert(user, "Activated")
```
Comment (MrMelbert) [inline code/game/machinery/teleporter.dm:148]:
```suggestion
		user.balloon_alert(user, "activated")
```

File: code/game/machinery/teleporter.dm [hunk 3]
```diff
@@ -97,12 +97,67 @@
...
+		return
+	GLOB.active_syndicate_beacons -= src
+	user.balloon_alert(user, "De-activated")
```
Comment (MrMelbert) [inline code/game/machinery/teleporter.dm:151]:
```suggestion
	user.balloon_alert(user, "deactivated")
```

File: code/modules/art/paintings.dm [hunk 4]
```diff
@@ -503,6 +503,104 @@
...
+	var/integrity_compromised = FALSE
+
+/obj/structure/sign/painting/syndicate_teleporter/Initialize(mapload, dir, building)
+	. = ..()
+	RegisterSignal(src, COMSIG_ATOM_TAKE_DAMAGE, PROC_REF(on_damage_taken))
+	RegisterSignals(src, list(COMSIG_QDELETING,	COMSIG_MACHINERY_BROKEN, TELEPORTER_SET_TARGET), PROC_REF(remove_connections))
+
+/obj/structure/sign/painting/syndicate_teleporter/Destroy()
+	if(current_canvas)
+		current_canvas.forceMove(drop_location())
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:523]: No side effects in destroy. This should be `QDEL_NULL(current_canvas)`, and you should implement `/obj/structure/sign/painting/syndicate_teleporter/atom_deconstruct()` which knocks it off the wall

File: code/modules/art/paintings.dm [hunk 5]
```diff
@@ -503,6 +503,104 @@
...
+	if(teleport_target)
+		teleport_target = null
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:525]:
```suggestion
	teleport_target = null
```

File: code/modules/art/paintings.dm [hunk 6]
```diff
@@ -503,6 +503,104 @@
...
+		return SECONDARY_ATTACK_CALL_NORMAL
+
+	if(teleport_target)
+		remove_connections(src)
+		user.balloon_alert(user, "Target cleared")
+		update_appearance(UPDATE_ICON)
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+	var/set_target = tgui_input_list(user, "Where to launch to?", "Set Teleporter?", GLOB.active_syndicate_beacons)
+	if(!set_target)
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:546]:
```suggestion
	if(!set_target || !user.is_holding(src))
```

File: code/modules/art/paintings.dm [hunk 7]
```diff
@@ -503,6 +503,104 @@
...
+	return ..()
+
+/obj/structure/sign/painting/syndicate_teleporter/attack_hand(mob/user, list/modifiers)
+	if(!teleport_target || !ishuman(user) || !IS_TRAITOR(user) || integrity_compromised)
+		return ..()
+	var/mob/living/carbon/human/teleportee = user
+	do_teleport(teleportee, teleport_target)
+
+/obj/structure/sign/painting/syndicate_teleporter/attack_hand_secondary(mob/user, list/modifiers)
+	. = ..()
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:535]:
```suggestion
	. = ..()
	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
		return .
```

File: code/modules/art/paintings.dm [hunk 8]
```diff
@@ -503,6 +503,104 @@
...
+	damage_deflection = 20
+	///Where a human gets teleported to once they activate the teleporter, picks from a list of GLOB.active_syndicate_beacons
+	var/atom/teleport_target
+	///If we've taken enough damage, we become unusable
+	var/integrity_compromised = FALSE
+
+/obj/structure/sign/painting/syndicate_teleporter/Initialize(mapload, dir, building)
+	. = ..()
+	RegisterSignal(src, COMSIG_ATOM_TAKE_DAMAGE, PROC_REF(on_damage_taken))
+	RegisterSignals(src, list(COMSIG_QDELETING,	COMSIG_MACHINERY_BROKEN, TELEPORTER_SET_TARGET), PROC_REF(remove_connections))
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:519]:
```suggestion
	RegisterSignals(src, list(COMSIG_QDELETING, COMSIG_MACHINERY_BROKEN, TELEPORTER_SET_TARGET), PROC_REF(remove_connections))
```

File: code/game/machinery/teleporter.dm [hunk 9]
```diff
@@ -97,12 +97,67 @@
...
+	return_target = return_portal
+	update_appearance(UPDATE_ICON)
+
+///Removes any active return_target
+/obj/machinery/teleport/hub/syndicate/proc/remove_connections(datum/source)
+	SIGNAL_HANDLER
+	if(return_target)
+		var/atom/old_return_target = return_target
+		return_target = null
+		SEND_SIGNAL(old_return_target, TELEPORTER_SET_TARGET, src)
```
Comment (MrMelbert) [inline code/game/machinery/teleporter.dm:141]: `TELEPORTER_SET_TARGET` should follow signal convention

File: code/modules/art/paintings.dm [hunk 10]
```diff
@@ -503,6 +503,104 @@
...
+	if(integrity_compromised)
+		icon_state = "BrokenFrame"
+		return
+	if(teleport_target)
+		icon_state = "ActivatedFrame"
+	else
+		icon_state = initial(icon_state)
+
+/obj/structure/sign/painting/syndicate_teleporter/crowbar_act(mob/living/user, obj/item/tool)
+	. = ..()
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:565]:
```suggestion
```
Not doing anything iirc

File: code/modules/art/paintings.dm [hunk 11]
```diff
@@ -503,6 +503,104 @@
...
+	if(!IS_TRAITOR(user))
+		return
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:567]:
```suggestion
		return NONE
```

File: code/modules/art/paintings.dm [hunk 12]
```diff
@@ -503,6 +503,104 @@
...
+	if(integrity_compromised)
+		qdel(src)
+		return
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:570]:
```suggestion
		return ITEM_INTERACT_SUCCESS
```

File: code/modules/art/paintings.dm [hunk 13]
```diff
@@ -503,6 +503,104 @@
...
+	remove_from_wall(user)
+
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:572]:
```suggestion
	remove_from_wall(user)
	return ITEM_INTERACT_SUCCESS
```

File: code/modules/art/paintings.dm [hunk 14]
```diff
@@ -503,6 +503,104 @@
...
+/obj/structure/sign/painting/syndicate_teleporter/proc/on_damage_taken(datum/source, damage_amount)
+	SIGNAL_HANDLER
+
+	var/total = max_integrity - damage_amount
+	var/integrity = (atom_integrity / total) * 100
+	if(integrity > 50)
+		return
+
+	if(current_canvas)
+		current_canvas.forceMove(drop_location())
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:590]:
```suggestion
	current_canvas?.forceMove(drop_location())
```

File: code/modules/art/paintings.dm [hunk 15]
```diff
@@ -503,6 +503,104 @@
...
+	remove_connections(src)
+	update_appearance(UPDATE_ICON)
+
+///Sends a signal so that all relevant teleporter hubs can clear their reference to us
+/obj/structure/sign/painting/syndicate_teleporter/proc/remove_connections(datum/source)
+	SIGNAL_HANDLER
+	if(teleport_target)
+		var/atom/old_teleport_target = teleport_target
+		teleport_target = null
+		SEND_SIGNAL(old_teleport_target, PAINTING_CUT_CONNECTIONS, src)
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:601]: Should follow signal name conventions.

File: code/modules/art/paintings.dm (hunk 10 above)
Comment (Xander3359) [reply code/modules/art/paintings.dm:565]: That's fine

File: code/modules/art/paintings.dm [hunk 16]
```diff
@@ -503,6 +503,104 @@
...
+		remove_connections(src)
+		user.balloon_alert(user, "Target cleared")
+		update_appearance(UPDATE_ICON)
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+	var/set_target = tgui_input_list(user, "Where to launch to?", "Set Teleporter?", GLOB.active_syndicate_beacons)
+	if(!set_target)
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+	teleport_target = set_target
+	SEND_SIGNAL(teleport_target, PAINTING_SET_TARGET, src)
```
Comment (MrMelbert) [inline code/modules/art/paintings.dm:549]: Signal conventions, again. But also this is starting to not look like a good use of signals. I'll check again next review.

File: code/datums/ruins/lavaland.dm [hunk 17]
```diff
@@ -62,6 +62,8 @@
 	suffix = "lavaland_surface_syndicate_base1.dmm"
 	cost = 20
 	allow_duplicates = FALSE
+	// Debug
+	always_place = TRUE
```
Comment (MrMelbert) [inline code/datums/ruins/lavaland.dm:66]: 🥴

File: code/modules/art/paintings.dm (hunk 10 above)
Comment (MrMelbert) [reply code/modules/art/paintings.dm:565]: Normally but not for these procs. These behaviors are pretty contained.

File: code/datums/ruins/lavaland.dm (hunk 17 above)
Comment (Xander3359) [reply code/datums/ruins/lavaland.dm:66]: This'll be removed when I take it out of draft 😅

File: code/modules/art/paintings.dm (hunk 10 above)
Comment (MrMelbert) [reply code/modules/art/paintings.dm:565]: I guess if you really wanna keep it you would have to make it respect parent return value
```suggestion
	. = ..()
	if(. & ITEM_INTERACT_BLOCKING)
		return .
```

**Comment by starrm4nn** [comment]
> 1. Set up syndicate painting
> 
>     2. Free respawns

maybe the base syndies should get teleblocker implants so its not usable as a free respawn, I LOVE FREE RESPAWNS GOING ON STATION AND VALIDHUNTING/ANTAGGING

**Comment by Xander3359** [comment]
> > 1. Set up syndicate painting
> >    
> >    2. Free respawns
> 
> maybe the base syndies should get teleblocker implants so its not usable as a free respawn, I LOVE FREE RESPAWNS GOING ON STATION AND VALIDHUNTING/ANTAGGING

Currently, non-traitors can't use the teleporter at the base.
This means any lavaland syndie cannot go through.

**Comment by CRITAWAKETS** [comment]
> Currently, non-traitors can't use the teleporter at the base. This means any lavaland syndie cannot go through.

step 1) build a quantum pad or teleporter
step 2) done

**Comment by Addust** [comment]
> > Currently, non-traitors can't use the teleporter at the base. This means any lavaland syndie cannot go through.
> 
> step 1) build a quantum pad or teleporter since you can just bring that to the syndiebase now step 2) done

step 3: get fucking orbitally nuked by admins or bombed by the miners for being a little shit

**Comment by Addust** [comment]
also give the scientists curtains at least

**Comment by Xander3359** [comment]
Closing this because I'm re-using the code in another PR

**Pull request closed without merging**